src/Npgsql/NpgsqlConnection.cs

```diff
@@ -1258,21 +1430,122 @@ void CheckDisposed()
+        internal ValueTask<NpgsqlConnector> StartBindingScope(
+            ConnectorBindingScope scope, NpgsqlTimeout timeout, bool async, CancellationToken cancellationToken)
+        {
+            // If the connection is around bound at a higher scope, we do nothing (e.g. copy operation started
+            // within a transaction on a multiplexing connection).
+            // Note that if we're in an ambient transaction, that means we're already bound and so we do nothing here.
+            if (ConnectorBindingScope != ConnectorBindingScope.None)
+            {
+                Debug.Assert(Connector != null, $"Connection bound with scope {ConnectorBindingScope} but has no connector");
+                Debug.Assert(scope != ConnectorBindingScope, $"Binding scopes aren't reentrant ({ConnectorBindingScope})");
+                return new ValueTask<NpgsqlConnector>(Connector);
+            }
+
+            return StartBindingScopeAsync();
+
+            async ValueTask<NpgsqlConnector> StartBindingScopeAsync()
+            {
+                Debug.Assert(Settings.Multiplexing);
+                Debug.Assert(_pool != null);
+
+                var connector = await _pool.Rent(this, timeout, async, cancellationToken);
+                ConnectorBindingScope = scope;
+                return connector;
+            }
+        }
+
+        internal NpgsqlConnector StartBindingScope(ConnectorBindingScope scope)
+            => StartBindingScope(scope, NpgsqlTimeout.Infinite, async: false, CancellationToken.None)
+                .GetAwaiter().GetResult();
+
+        internal EndScopeDisposable StartTemporaryBindingScope(out NpgsqlConnector connector)
+        {
+            connector = StartBindingScope(ConnectorBindingScope.Temporary);
+            return new EndScopeDisposable(this);
+        }
+
+        internal T CheckOpenAndRunInTemporaryScope<T>(Func<NpgsqlConnector, T> func)
+        {
+            CheckOpen();
+
+            using var _ = StartTemporaryBindingScope(out var connector);
+            var result = func(connector);
+            return result;
+        }
+
+        internal void EndBindingScope(ConnectorBindingScope scope)
+        {
+            Debug.Assert(ConnectorBindingScope != ConnectorBindingScope.None, $"Ending binding scope {scope} but connection's scope is null");
+
+            if (scope != ConnectorBindingScope)
+                return;
+
+            Debug.Assert(Connector != null, $"Ending binding scope {scope} but connector is null");
+            Debug.Assert(_pool != null, $"Ending binding scope {scope} but _pool is null");
+            Debug.Assert(Settings.Multiplexing, $"Ending binding scope {scope} but multiplexing is disabled");
+
+            // TODO: If enlisted transaction scope is still active, need to AddPendingEnlistedConnector, just like Close
+            var connector = Connector;
+            Connector = null;
+            connector.Connection = null;
+            _pool.Return(connector);
```

Comment:
I'm definitely not sure this is bug-free (not by a long-shot), but do you have any concrete scenario in mind?

Note that the design here is to only record the top-most binding scope, i.e. the one that caused the connection to bind to a connector. So if a transaction is started multiplexed connection, the binding scope is transaction. If after that a binding scope is started again (e.g. COPY operation), we ignore it entirely. As far as I know, no scope is reentrant - you can't start a transaction within a transaction, or a COPY within a COPY - otherwise we'd indeed need a stack or something.

Reply:
>As far as I know, no scope is reentrant

Good, that's the important bit, the rest will show itself in time :)